Update the docs of Slim FATE (#632)
FederatedAI/KubeFATE: docs/Introduction_to_Backend_Architecture.md

> # KubeFATE 部署FATE支持后端介绍

Reviewer: Will there be an English version for this doc?

Reviewer: If this is not plan to be done in this pr, then we need to create another issue to track.

Author: I will append an English document

FATE支持的计算引擎eggroll和spark

## 支持后端

FATE支持选择不同的计算引擎，存储引擎和数据传输引擎，现在KubeFATE支持快捷的部署不同引擎组合的FATE集群。

目前支持的组合包括以下几种：

- Eggroll
- Spark + hdfs + RabbitMQ
- Spark + hdfs + Pulsar
- Spark-local + localfs + Pulsar

### Eggroll

Eggroll是FATE支持计算存储引擎，包括rollsite，nodemanager，clustermanager三个组件

rollsite负责数据传输，nodemanager负责存储和计算，clustermanager负责管理。

docker-compose使用的时候修改`parties.conf`配置

```bash
backend=eggroll
```

k8s使用的时候修改`cluster.yaml`配置

```yaml
backend: eggroll
```

架构图：

<div align="center">
  <img src="./images/arch_eggroll.png" />
</div>

### spark_rabbitmq

> 

Reviewer: We need some buffering between chapters, like:
What is the drawback of eggroll and why we need rabbitmq.

当backend使用spark_rabbitmq的时候，会部署spark + hdfs + rabbitmq的引擎组合的FATE集群。

spark是计算组件，hdfs是存储组件，rabbitmq是数据传输组件。

docker-compose使用的时候修改`parties.conf`配置

```bash
backend=spark_rabbitmq
```

k8s使用的时候修改`cluster.yaml`配置

```yaml
backend: spark_rabbitmq
```

架构图：

<div align="center">
  <img src="./images/arch_spark_rabbitmq.png">
</div>

### spark_pulsar

> 

Reviewer: Add "why rabbitmq can be bad sometimes and why we need pulsar"

当backend使用spark_pulsar的时候，会部署Spark + Hdfs + Pulsar的引擎组合的FATE集群。

Spark是计算组件，Hdfs是存储组件，Pulsar是数据传输组件。

docker-compose使用的时候修改`parties.conf`配置

```bash
backend=spark_pulsar
```

k8s使用的时候修改`cluster.yaml`配置

```yaml
backend: spark_pulsar
```

架构图：

<div align="center">
  <img src="./images/arch_spark_pulsar.png">
</div>

### spark_local_pulsar (slim FATE)

> 

Reviewer: Now, why we need spark local pulsar?

# Introducing KubeFATE to deploy FATE-backed backends

> FATE supports eggroll and spark two computing engines

Reviewer: ```suggestion
FATE supports two computing engines: eggroll and spark
```

## Supported Backends

> FATE supports the selection of different computing engines, storage engines and federation engines, and now KubeFATE supports quick deployment of FATE clusters with different engine combinations.

Reviewer: ```suggestion
FATE supports multiple computing engines, storage engines and federation engines, and KubeFATE supports quick deployment of FATE clusters with different combination of engines.
```

List of supported engine combinations:

- **Eggroll**
- **Spark + hdfs + RabbitMQ**
- **Spark + hdfs + Pulsar**
- **Spark-local + localfs + Pulsar**

### Eggroll

> Eggroll is a computing storage engine natively supported by FATE, including three components: rollsite, nodemanager, and clustermanager

Reviewer: ```suggestion
Eggroll is a computing storage engine natively supported by FATE, including three components: rollsite, nodemanager, and clustermanager.
```

Rollsite is responsible for data transmission, nodemanager is responsible for storage and computing, and clustermanager is responsible for managing nodemanager.

Modify the `parties.conf` configuration when using docker-compose

```bash
backend=eggroll
```

> Modify the `parties.conf` configuration when using Kubernetes

Reviewer: ```suggestion
Modify the `parties.conf` configuration when using Kubernetes:
```

```yaml
backend: eggroll
```

Architecture diagram:

<div align="center">
  <img src="./images/arch_eggroll.png" />
</div>

### spark_rabbitmq

When backend uses spark_rabbitmq, a FATE cluster of spark + hdfs + rabbitmq engine combination will be deployed.

spark is the computing component, hdfs is the storage component, rabbitmq is the federation component.

> Modify the `parties.conf` configuration when using docker-compose

Reviewer: ```suggestion
Modify the `parties.conf` configuration when using docker-compose:
```

```bash
backend=spark_rabbitmq
```

> Modify the `parties.conf` configuration when using Kubernetes

Reviewer: ```suggestion
Modify the `parties.conf` configuration when using Kubernetes:
```

repository: <https://github.com/WeBankFinTech/eggroll.git>

Rabbitmq is a simple and easy-to-use Message Queue. It developed earlier and has more cloud platform support. It is suitable for federated learning with FATE of the Spark computing engine.

Compared with Rabbitmq, Pulsar can support larger-scale cluster deployment and also support exchange mode network structure. Using clustered deployment, it is suitable for larger-scale federated learning computing.

Compared with other modes, spark_local_pulsar can use this mode to minimize the components required for the cluster. It can also run FATE clusters with less resources and more resources, and can be used in small-scale federated learning computing, IOT devices, etc.

## Exchange Architecture

When using rollsite and Pulsar to federation, a federated learning network can be deployed using the Exchange architecture.

That is to say, when the backend is eggroll, spark_pulsar and spark_local_pulsar, it can support the use of Exchange.

**Through exchange, rollsite can only link with FATE of other rollsites, and Pulsar can only link with FATE of other Pulsar.**

> Reference documentation:

Reviewer: ```suggestion
Reference:
```
